src/modules/job-manager/raise.c

```diff
@@ -63,41 +65,61 @@ int raise_check_severity (int severity)
     return 0;
 }
 
-/* NB: job object may be destroyed in event_job_post_pack().
- * Do not reference the object after calling this function.
- * Do not call this function and continue to iterate on the job hash
- * with zhash_next().
- */
-int raise_job (struct job_manager *ctx,
-               const char *type,
-               int severity,
-               uint32_t userid,
-               const char *note,
-               struct job *job)
+int raise_job_exception (struct job_manager *ctx,
+                         struct job *job,
+                         const char *type,
+                         int severity,
+                         uint32_t userid,
+                         const char *note)
 {
     flux_jobid_t id = job->id;
     flux_future_t *f;
+    json_t *evctx;
 
-    if (event_job_post_pack (ctx->event,
-                             job,
-                             "exception",
-                             0,
-                             "{ s:s s:i s:i s:s }",
-                             "type", type,
-                             "severity", severity,
-                             "userid", userid,
-                             "note", note ? note : "") < 0)
-        return -1;
+    // create exception event context with the required keys per RFC 21
+    if (!(evctx = json_pack ("{s:s s:i}", "type", type, "severity", severity)))
+        goto nomem;
+    // work around flux-framework/flux-core#5314
+    if (!note)
+        note = "";
+    // add optional userid key
+    if (userid != FLUX_USERID_UNKNOWN) {
+        json_t *val;
+        if (!(val = json_integer (userid))
+            || json_object_set_new (evctx, "userid", val) < 0) {
+            json_decref (val);
+            goto nomem;
+        }
+    }
+    // add optional note key
+    if (note) {
+        json_t *val;
+        if (!(val = json_string (note))
+            || json_object_set_new (evctx, "note", val) < 0) {
+            json_decref (val);
+            goto nomem;
+        }
+    }
```

Comment:
given workaround above, note is always non-null ... or just checking `if (note)` for future expected changes

Reply:
Yes.  I figured I'd make it easy to just drop those two lines once the issue is addressed.